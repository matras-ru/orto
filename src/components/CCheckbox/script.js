import { selfInstall } from '@/utils';
import commonAttributes from '@/mixins/commonAttributes.js';

import ThemeClass from '@/themes/default/CCheckbox';

const {
    disabledClass,
    baseIconClass,
    defaultIconClass,
    checkedIconClass,
    disabledIconClass
} = ThemeClass;

export default {
    name: 'Checkbox',

    install(Vue, theme) {
        selfInstall(Vue, theme, this);
    },

    props: {
        ...commonAttributes.props,
        label: {
            type: String,
            default: null
        },
        hint: {
            type: String,
            default: null
        },
        modelValue: {
            type: [Array, Boolean],
            default: null
        },
        value: {
            type: [String, Number],
            default: null
        },
        trueValue: {
            type: Boolean,
            default: true
        },
        falseValue: {
            type: Boolean,
            default: false
        }
    },

    model: {
        prop: 'modelValue',
        event: 'change'
    },

    data() {
        return {
            isChecked: false
        };
    },

    computed: {
        currentClass() {
            let classes = ['flex mb-1-4'];

            if (this.disabled) {
                classes.push(disabledClass);
            }

            return classes;
        },
        currentIconClass() {
            let classes = [baseIconClass];

            if (this.disabled) {
                classes.push(disabledIconClass);
            } else if (this.isChecked || this.shouldBeChecked) {
                classes.push(checkedIconClass);
            } else {
                classes.push(defaultIconClass);
            }

            return classes;
        },
        shouldBeChecked() {
            if (this.modelValue instanceof Array) {
                return this.modelValue.includes(this.value);
            }

            return this.modelValue === this.trueValue;
        }
    },

    methods: {
        onChange(e) {
            this.isChecked = e.target.checked;

            if (this.modelValue instanceof Array) {
                let newValue = [...this.modelValue];

                if (this.isChecked) {
                    newValue.push(this.value);
                } else {
                    newValue.splice(newValue.indexOf(this.value), 1);
                }

                this.$emit('change', newValue);
            } else {
                this.$emit('change', this.isChecked ? this.trueValue : this.falseValue);
            }
        }
    },

    render(h) {
        const attrs = {
            id: this.id,
            autofocus: this.autofocus,
            name: this.name,
            type: 'checkbox',
            label: this.label,
            value: this.value,
            disabled: this.disabled
        };

        const on = {
            change: this.onChange
        };

        const componentData = {
            class: 'absolute opacity-0 pointer-events-none',
            attrs: attrs,
            on
        };

        return h('label', { class: this.currentClass }, [
            h('input', componentData),
            h('div', { class: 'flex' }, [
                h('span', { class: this.currentIconClass }),
                h('span', this.label)
            ])
        ]);
    }
};
