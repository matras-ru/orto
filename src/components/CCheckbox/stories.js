import { storiesOf } from '@storybook/vue';
import { action } from '@storybook/addon-actions';
import { withKnobs, text, boolean } from '@storybook/addon-knobs';

import CCheckbox from './';

const stories = storiesOf('CCheckbox', module);
stories.addDecorator(withKnobs);

stories
    .add('Checkbox knobs', () => ({
        components: { CCheckbox },
        props: {
            id: {
                default: text('id', 'checkbox1')
            },
            autofocus: {
                default: boolean('autofocus', false)
            },
            label: {
                default: text('label', 'Checkbox1')
            },
            name: {
                default: text('name', 'checkbox1')
            },
            disabled: {
                default: boolean('disabled', false)
            },
            value: {
                default: text('value', 'checkbox1')
            }
        },
        data() {
            return {
                checkboxModel: false
            };
        },
        template:
            '<CCheckbox @onChange="action" :id="id" :autofocus="autofocus" :name="name" :label="label" :disabled="disabled" :value="value" v-model="checkboxModel" />',
        methods: { action: action('change') }
    }))
    .add('Default checkbox', () => ({
        components: { CCheckbox },
        data() {
            return {
                checkboxModel: false
            };
        },
        template:
            '<CCheckbox @onChanget="action" id="checkbox2" label="Checkbox2" name="checkbox2" value="checkbox2" v-model="checkboxModel" />',
        methods: { action: action('change') }
    }))
    .add('Disabled checkbox', () => ({
        components: { CCheckbox },
        template:
            '<CCheckbox @onChange="action" id="checkbox3" label="Checkbox3" name="checkbox3" value="checkbox3" disabled />',
        methods: { action: action('change') }
    }));
