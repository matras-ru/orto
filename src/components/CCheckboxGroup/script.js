import { selfInstall } from '@/utils';
import CCheckbox from '@/components/CCheckbox';

export default {
    name: 'CheckboxGroup',

    install(Vue, theme) {
        selfInstall(Vue, theme, this);
    },

    components: {
        CCheckbox
    },

    props: {
        data: {
            type: Array,
            default: () => []
        },
        modelValue: {
            type: Array,
            default: () => []
        }
    },

    model: {
        prop: 'modelValue',
        event: 'change'
    },

    computed: {
        values: {
            get() {
                return this.modelValue;
            },
            set(val) {
                this.$emit('change', val);
            }
        }
    },

    render(h) {
        let setProps = item => {
            return {
                id: item.id,
                autofocus: item.autofocus,
                label: item.label,
                name: item.name,
                value: item.value,
                disabled: item.disabled
            };
        };

        return h(
            'div',
            this.data.map(item => {
                return h('CCheckbox', {
                    props: {
                        modelValue: this.values,
                        ...setProps(item)
                    },
                    on: {
                        change: arr => {
                            this.values = [...arr];
                        }
                    }
                });
            })
        );
    }
};
