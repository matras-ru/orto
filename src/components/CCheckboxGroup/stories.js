import { storiesOf } from '@storybook/vue';

import CCheckboxGroup from './';

const stories = storiesOf('CCheckboxGroup', module);

stories.add('CCheckboxGroup', () => ({
    components: { CCheckboxGroup },
    template:
        "<CCheckboxGroup :data=\"[{ id: 'checkbox4', label: 'Checkbox1', name: 'checkbox4' }, { id: 'checkbox5', label: 'Checkbox2', name: 'checkbox5' }]\" />"
}));
