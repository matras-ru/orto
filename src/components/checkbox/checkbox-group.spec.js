/* eslint-env jest */
import { mount, config, createLocalVue } from '@vue/test-utils';

import CCheckboxGroup from './checkbox-group';
import CCheckbox from './checkbox';

config.stubs['CCheckbox'] = CCheckbox;

describe('Checkbox group', () => {
    it('is rendered & not empty', () => {
        const wrapper = mount(CCheckboxGroup, {
            context: {
                props: {
                    modelValue: [],
                    data: [
                        {
                            id: 'chkb1',
                            label: 'chkb1',
                            name: 'chkb1',
                            value: 'chkb1'
                        },
                        {
                            id: 'chkb2',
                            label: 'chkb2',
                            name: 'chkb2',
                            value: 'chkb2'
                        }
                    ]
                }
            }
        });

        expect(wrapper.findAll('input[type="checkbox"]').length).toBe(2);
    });

    it('v-model', () => {
        const localVue = new createLocalVue();

        const App = localVue.extend({
            data() {
                return {
                    model: []
                };
            },

            render(h) {
                return h(CCheckboxGroup, {
                    props: {
                        data: [
                            {
                                id: 'chkb1',
                                label: 'chkb1',
                                name: 'chkb1',
                                value: 'chkb1'
                            },
                            {
                                id: 'chkb2',
                                label: 'chkb2',
                                name: 'chkb2',
                                value: 'chkb2'
                            }
                        ],
                        modelValue: this.model
                    },
                    on: {
                        change: val => {
                            this.model = val;
                        }
                    }
                });
            }
        });

        const wrapper = mount(App, {
            localVue: localVue
        });

        const input = wrapper.find('input[type="checkbox"]');

        expect(wrapper.vm.model).toEqual([]);
        input.setChecked();
        expect(wrapper.vm.model).toEqual(['chkb1']);

        wrapper.destroy();
    });
});
