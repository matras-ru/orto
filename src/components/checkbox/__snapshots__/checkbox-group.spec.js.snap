// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox group is rendered & not empty 1`] = `
<div class="mb-2-2">
  <div>
    <div>
      <div class="flex flex-wrap mb-0-4"><label for="chkb1" class="relative pl-1-4 cursor-pointer"><input id="chkb1" name="chkb1" type="checkbox" class="absolute top-0-2 left-0 w-0-8 h-0-8 form-checkbox" value="chkb1">chkb1</label></div>
      <div class="flex flex-wrap mb-0-4"><label for="chkb2" class="relative pl-1-4 cursor-pointer"><input id="chkb2" name="chkb2" type="checkbox" class="absolute top-0-2 left-0 w-0-8 h-0-8 form-checkbox" value="chkb2">chkb2</label></div>
    </div>
  </div>
</div>
`;
